packages/lib/downloadController.ts

```diff
@@ -0,0 +1,114 @@
+import Logger from '@joplin/utils/Logger';
+import JoplinError from './JoplinError';
+import { ErrorCode } from './errors';
+
+const logger = Logger.create('downloadController');
+
+export interface DownloadController {
+	totalBytes: number;
+	imagesCount: number;
+	imageCountExpected: number;
+	printStats(imagesCountExpected: number): void;
+	handleNewChunk(request: any): (chunk: any)=> void;
+	limitMessage(): string;
+}
+
+export class DummyDownloadController implements DownloadController {
+	public totalBytes = 0;
+	public imagesCount = 0;
+	public imageCountExpected = 0;
+	public printStats(): void {}
+	public handleNewChunk() { return () => {}; }
+	public limitMessage() { return ''; }
+}
+
+export class LimitedDownloadController implements DownloadController {
+	private totalBytes_ = 0;
+	// counts before the downloaded has finished, so at the end if the totalBytes > maxTotalBytesAllowed
+	// it means that imageCount will be higher than the total download during the process
+	private imagesCount_ = 0;
+	// how many images links the email has
+	private imageCountExpected_ = 0;
+	private isLimitExceeded_ = false;
+
+	private maxTotalBytes = 0;
+	private maxImagesCount = 0;
+	private ownerId = '';
+
+	public constructor(ownerId: string, maxTotalBytes: number, maxImagesCount: number) {
+		this.ownerId = ownerId;
+		this.maxTotalBytes = maxTotalBytes;
+		this.maxImagesCount = maxImagesCount;
+	}
+
+	public set totalBytes(value: number) {
+		if (this.totalBytes_ >= this.maxTotalBytes) {
+			throw new JoplinError(`Total bytes stored (${this.totalBytes_}) has exceeded the amount established (${this.maxTotalBytes})`, ErrorCode.DownloadLimiter);
+		}
+		this.totalBytes_ = value;
+	}
+
+	public get totalBytes() {
+		return this.totalBytes_;
+	}
+
+	public set imagesCount(value: number) {
+		if (this.imagesCount_ > this.maxImagesCount) {
+			throw new JoplinError(`Total images to be stored (${this.imagesCount_}) has exceeded the amount established (${this.maxImagesCount})`, ErrorCode.DownloadLimiter);
+		}
+		this.imagesCount_ = value;
+	}
+
+	public get imagesCount() {
+		return this.imagesCount_;
+	}
+
+	public set imageCountExpected(value: number) {
+		this.imageCountExpected_ = value;
+	}
+
+	public get imageCountExpected() {
+		return this.imageCountExpected_;
+	}
+
+	public handleNewChunk(request: any) {
+		return (chunk: any) => {
+			try {
+				this.totalBytes += chunk.length;
+			} catch (error) {
+				request.destroy(error);
```

Comment:
Don't catch the error here - let the handle throw and the request object should handle it properly and run any cleanup

Reply:
The error we are catching here is from the `totalBytes` setter. I can test again, but the last time I tested the error was crashing the server (which seems a little strange, to be honest, I'm not sure if I was doing something wrong).

Maybe it crashes because it is being called from a callback?
https://github.com/laurent22/joplin/blob/d602b729eadddb6566507c7b5267827419032972/packages/lib/shim-init-node.ts#L553-L580